### PR TITLE
Fix comparisons for array fields

### LINKED ISSUE #7615: Compiler fails to generate default comparison between records with array fields
### Summary of Problem
The Chapel compiler generates incorrect code for comparison between
records that have fields that are arrays.

### Steps to Reproduce

**Source Code:**
```chpl
record R {
  var A : [1..5] int;
}

var A, B : R;
writeln(A == B);
```

```
In file included from /path/to/gen/_main.c:43:0:
/path/to/gen/foo.c: In function ‘chpl___EQUALS__2’:
/path/to/gen/foo.c:146:5: error: used struct type value where scalar is required
 if (call_tmp_chpl) {
     ^~~~~~~~~~~~~
gmake: *** [/path/to/gen/foo.tmp] Error 1
error: compiling generated source
```

**Compile command:**
`chpl foo.chpl`

**Associated Future Test(s):**
[types/records/bharshbarger/recordArrCmp.chpl](https://github.com/chapel-lang/chapel/blob/master/test/types/records/bharshbarger/recordArrCmp.chpl)
[types/records/bharshbarger/recordDomArrCmp.chpl](https://github.com/chapel-lang/chapel/blob/master/test/types/records/bharshbarger/recordDomArrCmp.chpl)

### Configuration Information

All configurations, present since version 1.7

### LINKED ISSUE #11487: Associative Domains with Records containing Arrays as keys causes non-deterministic compiler crash
```chpl
record R {
	var a : [1..10] int;
}
var d : domain(R);
d += new R();

writeln(d);
```

[TIO](https://tio.run/##S85ILEjN@f@/KDU5vyhFIUihmouzLLFIIVHBSiHaUE/P0CBWITOvxJqrlgsknAIUTsnPTczM0wjStOZKUdC2VchLLVcI0gDyuMqLMktSc/I0UjSt//8HAA)

You may need to rerun this a few times.

### LINKED ISSUE #16821: HashedDist fails when dmapped across types not defined with by name.
### Summary of Problem

I am trying to hash my own custom records onto the locales. I have set up a record, called `State`. The challenge is to define a [HashedDist](https://chapel-lang.org/docs/modules/dists/HashedDist.html) across these states.

However HashedDist crashes the compiler when the input type is not explicitly set.
the problematic line is `      const D: domain(this.eltType) dmapped Hashed(idxType=this.eltType);`

To give you some context I am solving the board game connect 4.

### Output
```bash
chpl POC.chpl 
internal error: UTI-MIS-0788 chpl version 1.23.0

Internal errors indicate a bug in the Chapel compiler ("It's us, not you"),
and we're sorry for the hassle.  We would appreciate your reporting this bug -- 
please see https://chapel-lang.org/bugs.html for instructions.  In the meantime,
the filename + line number above may be useful in working around the issue.
```

```bash
chpl POC.chpl --devel
internal error: seg fault [util/misc.cpp:788]
```

### Steps to Reproduce
`chpl POC.chpl` the source code below.

**Source Code:**

save this to POC.chpl below
```chapel

use HashedDist;

enum Player {
  Red,
  Yellow
}

enum Tile {
  Red,
  Yellow,
  Unset
}

const numRows : int = 6;
const numColumns : int = 7;
const DLocal : domain(2) = {0..numColumns, 0..numRows};

record State {
  var player : Player;
  var board : [DLocal] Tile;

  proc init() {
  }

  proc init(player : Player, board) {
    this.player = player;
    this.board = board;
  }
}


public class Searcher {
    type eltType;
    
    
    proc init(type eltType) {
      this.eltType = eltType;
      this.complete();
    }

    /* search function.
    */
    proc search(start : this.eltType) {
      const D: domain(this.eltType) dmapped Hashed(idxType=this.eltType);
      var fScore : [D] real;
      var gScore : [D] real;
    }
}

proc main() {
    const startState = new State();
    var searcher = new Searcher(startState.type);
    searcher.search(startState);
}
```

**Compile command:**
`chpl POC.chpl`

**Execution command:**
N/A

**Associated Future Test(s):**
N/A

### Configuration Information

- Output of `chpl --version`:

```bash
chpl version 1.23.0
Copyright 2020 Hewlett Packard Enterprise Development LP
Copyright 2004-2019 Cray Inc.
(See LICENSE file for more details)
```

- Output of `$CHPL_HOME/util/printchplenv --anonymize`:

```bash
CHPL_TARGET_PLATFORM: darwin
CHPL_TARGET_COMPILER: clang
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: gmp
CHPL_HWLOC: hwloc
CHPL_REGEXP: re2
CHPL_LLVM: none
CHPL_AUX_FILESYS: none
```

- Back-end compiler and version, e.g. `gcc --version` or `clang --version`:
```bash
gcc --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/4.2.1
Apple clang version 12.0.0 (clang-1200.0.32.27)
Target: x86_64-apple-darwin20.1.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

```bash
clang --version
Apple clang version 12.0.0 (clang-1200.0.32.27)
Target: x86_64-apple-darwin20.1.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

- (For Cray systems only) Output of `module list`:
N/A

### PR DESCRIPTION
**Pull request by bradcray**
Currently, compiler-generated comparisons (`==`, `!=`, etc.) for records containing array fields result in compile-time accesses to freed memory.  This PR attempts to improve the situation by:

* for compiler-generated `==` and `!=` operators on records, implementing a helper function, `chpl_field_neq`, with array and non-array overloads, where the array overloads apply a reduction to the promoted `!=` operator and the non-array cases just call `!=`
* for compiler-generated `<`, `<=`, `>`, and `>=`, implementing helpers `chpl_field_lt` and `chpl_field_gt` which generate compiler errors saying that array fields don't support ordered comparisons by default, and calling the normal operators for traditional fields.

Beyond this, I:

* retired the futures related to #7615.
* added a new test for the ordered comparison cases
* added the test case from #11487

This PR also contains a reasonably inconsequential (and ultimately desirable) change to DefaultSparse.chpl that I'd been carrying around in my repo and committed accidentally.  I can pull it out of this PR if people prefer (though eventually, I think it belongs on master).

Resolves #7615.
Resolves #11487.
Resolves #16821

### CODE REVIEW

**Comment by bradcray** [comment]
This is the PR that fixes the bug I demo'd today, though I'm realizing belatedly that it has one failure for which I'm looking for advice on from @e-kayrakli.  Specifically, https://github.com/chapel-lang/chapel/blob/master/test/types/records/engin/comparison.chpl fails due to its definition of user-defined overloads of operators on classes.  I believe that this is due to the new point-of-instantiation rules which say that the `==` operators in my new `ChapelBase.chpl` routines will be evaluated from their scope rather than the callees' scopes, which means that they're using the normal "pointer ==" definitions.

I was originally thinking that this meant my approach was broken (at least until we implement such operators as methods as currently proposed which would fix the POI issues, I think?).  However, in typing this up, I'm feeling unsure whether users are supposed to be able to over-ride `==` and `!=` for classes, given that they can't override `=` (I'm even less sure about `<`, `>`, `<=`, `>=`...?), making me wonder whether what the test is doing is even reasonable?

If the test _is_ legal then I need a way to work around the POI problem.  So far, the only ones I've come up with are (a) add the support for type method operator overloads to the  language and then use them for this test; (b) put the new logic that handles array fields into the compiler-generated routine itself (which has the downsides of potentially adding unnecessary cases for each field); (c) have the compiler generate non-generic `chpl_field_eq`, `chpl_field_neq`, etc. calls for each user-defined class or record next to that record so that any operators defined in the type's scope are visible; (d) something else?

Tagging @vasslitvinov as well due to his POI experience and @lydia-duncan due to the relationship with operators.

**Comment by bradcray** [comment]
Here's an issue that (mostly) asserts that if we don't support user-defined `=` on classes, we shouldn't support user-defined `==` or `!=` either because it causes problems for comparing against `nil`, disables our most primitive way for checking a class's value, etc.  https://github.com/chapel-lang/chapel/issues/13048

**Comment by bradcray** [comment]
Here's where this PR stands:

* If we continue to support user-defined `==` and `!=` on classes (which is how things currently seem to be trending), then we need a way to keep `test/types/records/engin/comparison.chpl` working.  I think the sanest way to do this would be to complete the plan to implement operators as [type] methods which would permit a user-defined `==` to be visible within the utility functions I added to ChapelBase.chpl, where today our POI rules don't permit it to be.  Alternatives would be to move away from implementing the field comparison utilities in ChapelBase.chpl, for example by injecting them into the modules that define the types themselves.  However, that feels clunkier and less elegant to me.

* If we cease to support user-defined `==` and `!=` on classes (as in PR #16832) then that test will need to be rewritten and the current approach will continue to work.

**Comment by bradcray** [comment]
@e-kayrakli / @mppf / @lydia-duncan / @vasslitvinov : I'm looking for guidance about whether we should consider merging this bug fix for the release or not.  The quick summary is:

* default compiler-generated comparisons for records with array fields go off the rails in weird ways, and (painfully) have since the dawn of time.  This was encountered by a user recently, and took some time to track down because the failure mode is pretty unclear.  The workaround when encountering this bug is to have the user define their own `==` overload for their record type.

* I think the fix in this PR is pretty straightforward.  However due to POI/operator method issues, it breaks an existing test of Engin's, which is among the few test cases we have which (a) defines a user-defined `==` operator for classes and (b) relies on a compiler-generated `==` function for a record with a field of that class type.  As a result, it uses the default `==` operator on classes rather than Engin's overload.

* We've discussed the likely ideal way forward for this PR to be to redefine `==` operators for classes as methods and to have Engin's test use a `==` operator method rather than a standalone function.  However, that does not yet work as expected due to issue #16853, which doesn't look likely to be fixed by this release.

So our choices are essentially between:

* sticking with the status quo, where array fields break compiler-generated record comparisons

* merging this PR, which fixes compiler-generated record comparisons for array fields, but will not call user-defined `==` operators on class fields (splitting Engin's test into "working" and "not yet working" versions)

Arguing either side:

Status quo:
* no worse than we were before

Merging:
* fixes an issue users have run into
* we've historically been vague/uncertain about whether users can define `==` on classes, and it's not clear many do.  However, Arkouda does.

I think based on this, we should probably stick with the status quo since it's more conservative, but wanted to get additional opinions rather than making that call myself.

**Comment by mppf** [comment]
My view: I'd be OK either way. If we want to merge it we should make sure Arkouda still works with it first.

**Comment by e-kayrakli** [comment]
Can we add an error for record comparisons if the record has an array field and doesn't have an overload of the comparison operator? We can indicate that it is an unimplemented feature, and the user should add their own overload. I think that'd be the ideal solution today.

In general, reading this issue, I was ready to say the test cases in that test is not important and given the vagueness around `==` for classes, we can drop that part of the test or split it into a future or something. But reading that Arkouda has an overload changed my mind and wanted me to keep the current status in terms of behavior.

**Comment by bradcray** [comment]
> Can we add an error for record comparisons if the record has an array field and doesn't have an overload of the comparison operator? We can indicate that it is an unimplemented feature, and the user should add their own overload. I think that'd be the ideal solution today.

I think we could, but I didn't have time this week.

**Comment by bradcray** [comment]
Noting for my own sake:  This PR is blocked by https://github.com/Cray/chapel-private/issues/1731 but I haven't had any luck expressing this as a cross-repo dependency, perhaps because the other repo isn't public.

File: modules/internal/ChapelBase.chpl [hunk 1]
```diff
@@ -2372,4 +2372,31 @@ module ChapelBase {
   inline proc _removed_cast(in x) {
     return x;
   }
+
+  /* The following two overloads support compiler-generated == and !=
+     operators for records with array fields */
```
Comment (e-kayrakli) [inline modules/internal/ChapelBase.chpl:2377]: You might want to make this comment more general to encompass the other four, or add similar comments to those

File: modules/internal/DefaultSparse.chpl [hunk 2]
```diff
@@ -49,6 +49,10 @@ module DefaultSparse {
       this.dist = dist;
     }
 
+    proc stridable param {
+      return parentDom.stridable;
+    }
+
```
Comment (e-kayrakli) [inline modules/internal/DefaultSparse.chpl:55]: I don't mind if this goes in on this PR. The previous behavior was... that `stridable` was unresolved on DefaultSparse?

**Review by e-kayrakli — APPROVED** [review]

File: modules/internal/ChapelBase.chpl (hunk 1 above)
Comment (bradcray) [reply modules/internal/ChapelBase.chpl]: Good point, I've fixed it.

File: modules/internal/DefaultSparse.chpl (hunk 2 above)
Comment (bradcray) [reply modules/internal/DefaultSparse.chpl:55]: Yes.  And I think this comes up in contexts (like slicing maybe?) where we are applying general recipes in ChapelArray.chpl to arbitrary array/domain implementations.

**Review by e-kayrakli — APPROVED** [review]

**Pull request merged**